NumPyInference Batching

nitron532 opened the pull request:
For issue #4 (Inference Batching).

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -201,21 +200,21 @@ def _run_layer(
-        This flattens the 2D image data into 1D, adds a batch dimension, and adjusts the
-        image data by ``input_scale`` and ``input_zero``.
+        This adjusts the batch image data by ``input_scale`` and ``input_zero``.
+        Then, it flattens the 2D image(s) into 1D and adds the batch dimension.
```
"flattens each 2D image into 1D" ?

"adds the batch dimension" seems inaccurate, since `test_batch` should already have a batch dimension?

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -201,21 +200,21 @@ def _run_layer(
+        Then, it flattens the 2D image(s) into 1D and adds the batch dimension.
 
-        :param test_image: Image data to preprocess. This data should have already been
+        :param test_batch: Batch data to preprocess. This data should have already been
```
The parameter should be renamed to `test_batch` consistently, the rest of the code still calls it `test_image`

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -201,21 +200,21 @@ def _run_layer(
 
-        :param test_image: Image data to preprocess. This data should have already been
+        :param test_batch: Batch data to preprocess. This data should have already been
             normalized to [0.0, 1.0] and resized to 12×12, usually by
```
This should also mention the shape of `test_batch` somewhere `(batch_size, 12, 12)`; it may not be obvious to the reader that the batch dimension is first

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -201,21 +200,21 @@ def _run_layer(
 
-        :returns: Flattened image data, adjusted by the quantized neural network's
+        :returns: Flattened batch data, adjusted by the quantized neural network's
                   ``input_scale`` and ``input_zero``.
```
Mention the shape of the returned data `(144, batch_size)`; not obvious that the batch dimension becomes last

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -230,32 +229,36 @@ def preprocess_image(self, test_image: np.ndarray) -> np.ndarray:
-        ).astype(np.int8)
+
+        #Flatten 2D -> 1D
+        return test_image.reshape(test_image.shape[0], -1).T
```
The rest of the code in this file prefers `.transpose()` over `.T`, so we should be consistent here

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -201,21 +200,21 @@ def _run_layer(
 
     def preprocess_image(self, test_image: np.ndarray) -> np.ndarray:
-        """Preprocess the raw image data, as required by the quantized neural network.
+        """Preprocess the raw image data in the batch.
```
`NumPyInference.preprocess_image` is also called by [`fpga/fpga_inference.py`](https://github.com/UCSBarchlab/pyrtlnet/blob/main/fpga/fpga_inference.py), which will also need to be updated.

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -230,32 +229,36 @@ def preprocess_image(self, test_image: np.ndarray) -> np.ndarray:
+                    with shape ``(10, batch_size)``.
+                    Note that these layer outputs are transposed
+                    compared to :func:`.run_tflite_model`.
+                    ``actuals`` is an np.ndarray of predicted digits.
```
Mention the shape of `actuals` `(batch_size,)`

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -230,32 +229,36 @@ def preprocess_image(self, test_image: np.ndarray) -> np.ndarray:
-        layer1_output = self._run_layer(
-            1, layer0_output, self.layer[0].zero, run_relu=False
-        )
+        # Convert single image to batchable shape
```
Is this logic needed? Seems better to simplify this code by assuming that the input already has a batch dimension. We could add an `assert` if you want to be sure that the input has the right shape.

nitron532 replied:
I added this one because the pytest file passed a single image (a standalone matrix) straight to self.run(), while my logic assumed every image (even if it was just one image like in pytest) was preprocessed into an array of matrices in the batch loop that calls self.run(). So basically I just added it to pass the test haha, I can remove it but it wont pass the numpy inference test unless we modify the way an image is passed into the model.

fdxmw commented on numpy_inference.py:
```diff
@@ -17,7 +17,9 @@ def main() -> None:
     parser = argparse.ArgumentParser(prog="numpy_inference.py")
     parser.add_argument("--start_image", type=int, default=0)
     parser.add_argument("--num_images", type=int, default=1)
+    parser.add_argument("--batch_size", type=int, default=1)
```
Add a `help=` that describes how `--batch_size` interacts with `--num_images`

fdxmw commented on numpy_inference.py:
```diff
@@ -32,46 +34,77 @@ def main() -> None:
+    if args.batch_size > args.num_images:
+        args.num_images = args.batch_size
+
+    if args.num_images > len(test_images):
```
This calculation should consider `start_image`, since that may reduce the effective number of available `test_images`

fdxmw commented on numpy_inference.py:
```diff
@@ -32,46 +34,77 @@ def main() -> None:
+        batch_end_index = min(batch_index + args.batch_size, len(test_images))
+        vanilla_batch = test_images[batch_index:batch_end_index]
+
+        test_batch = np.array(vanilla_batch)
```
I don't think you need this call to `np.array`?

fdxmw reviewed (commented):
Overall, your change good to me! Though I do have some comments that will need to be addressed.

Also, you will need to update `numpy_inference_test` to test your new batching logic before we can merge this PR.

Let me know if you have questions or comments!

fdxmw commented on pyrtlnet/numpy_inference.py:
```diff
@@ -230,32 +231,41 @@ def preprocess_image(self, test_image: np.ndarray) -> np.ndarray:
-                  is the actual predicted digit. ``predicted_digit`` is equivalent to
-                  ``layer1_output.flatten().argmax()``.
+        :param test_batch: A batch of shape (batch_size, 12, 12)
+        to run through the NumPy inference implementation.
```
This line needs to be indented 4 more spaces so Sphinx understands that it's a continuation of the `test_batch`'s description, rather than the start of a new paragraph. See the [generated documentation for your pull request](https://pyrtlnet--5.org.readthedocs.build/en/5/numpy_inference.html#pyrtlnet.numpy_inference.NumPyInference.run) and the other `:param:` directives in this file

nitron532 commented on fpga/fpga_inference.py:
```diff
@@ -80,14 +80,15 @@ def main() -> None:
     flat_batch = numpy_inference.preprocess_image(test_batch)
     # Convert the signed image data to raw byte values.
-    for flat_image in flat_batch.transpose():
+    for image_index in flat_batch.transpose():
```
oops... the linter was complaining that I was reassigning the iterative object, so I tried to use indexing to get the flat_image but I still messed up the indexing.

fdxmw replied:
Ah I see, maybe we could call the list comprehension output `flat_image_bytes` or similar?

nitron532 commented:
Ok, I just added a few things:
1. --verbose and --no-verbose flags now output similar descriptions, just that --no-verbose skips the layer outputs and images. There's a summary of each image and its relation to batch and test_image index like your suggestion, and the empty space print messages should be fixed now. I also corrected the indexing of the batch inputs to follow the global numbering (batch_start_index + batch_index).

2. The fpga_inference.py was renamed to flat_image_bytes in the list comprehension line, but I kept the indexing so the linter wouldn't yell at me for changing the iterative variable.

3. I fixed the doc indentation, I just checked the Read the Docs build it made and it looks right.

That should be it, I also just talked to Professor Sherwood a few hours ago, he said I can come by the ArchLab on Friday to pick up one of their FGPAs to test on. Pretty excited to do that!
As always let me know if there's anything more to be fixed on this PR!
